### ISSUE TITLE
[v1.3.4-evolve] transformer-relevant tensor AD returns wrong gradients or rejects dual input

### ISSUE THREAD

**Issue by Gabriel-Kahen** [tsotchke/eshkol#551]
Observed with clean `tsotchke/eshkol@90cbd7130f47b8184bcc77b8d5c1b0026da980de`, `Eshkol Compiler v1.3.4-evolve`, compiler SHA-256 `caa295b19a6e9388963aa0def99dade63656d2dcbffccad421bd1daaa1db3750` on Linux x86-64. Upstream HEAD was not rebuilt.

Repeated AOT/JIT probes found three transformer-relevant AD gaps while scalar, matmul, ReLU, SiLU, repeated-call, control-flow, and repeated-index embedding checks passed.

1. Generic tensor gradient returns values that fail `[4, 3]`:

```scheme
(define actual
  (gradient (lambda (v)
              (+ (* (tensor-get v 0) (tensor-get v 0))
                 (* 3.0 (tensor-get v 1))))
            #(2.0 5.0)))
(display actual) (newline)
```

2. The derivative of one scaled-dot-attention output with respect to one value element fails a central finite-difference check in all four runs:

```scheme
(define q (reshape #(0.0 0.0 0.0 0.0) 2 2))
(define k (reshape #(0.0 0.0 0.0 0.0) 2 2))
(define (f x)
  (tensor-get (scaled-dot-attention q k
                (reshape (vector x 2.0 3.0 4.0) 2 2)) 0 0))
(display (derivative f 1.0)) (newline)
```

3. Differentiating through `layer-norm` rejects the dual-bearing vector in both engines:

```scheme
(define (f x) (tensor-get (layer-norm (vector x 2.0 3.0) 1.0 0.0 1e-5) 0))
(display (derivative f 1.0)) (newline)
```

Diagnostic: `Type error in layer-norm: expected tensor or numeric vector, got vector`.

Expected: advertised differentiable tensor operations match finite differences/reference gradients, or unsupported point/operation combinations reject explicitly before returning a wrong gradient.

**Comment by Gabriel-Kahen** [comment]
N2 downstream reproducer/addendum on the same pinned revision.

Environment: clean `tsotchke/eshkol@90cbd7130f47b8184bcc77b8d5c1b0026da980de`,
`Eshkol Compiler v1.3.4-evolve`, compiler SHA-256
`f659f81dfa00864f1b2c16c0fc456aa7caa460a3b064aea9654dec1803f2e4bb`,
unsupported CachyOS x86-64 / LLVM-Clang 22.1.6 compatibility lane.

This tensor-input form does not reject the operation. It returns a zero reverse
gradient where the same compiled program's central difference is nonzero:

```scheme
(define (near a b tolerance) (< (abs (- a b)) tolerance))
(define (objective values)
  (tensor-get (layer-norm values 1.0 0.0 1e-5) 0))
(define point #(1.0 2.0 3.0))
(define step 1e-5)
(define actual (gradient objective point))
(define plus (objective #((+ 1.0 step) 2.0 3.0)))
(define minus (objective #((- 1.0 step) 2.0 3.0)))
(define finite (/ (- plus minus) (* 2.0 step)))
(display "reverse=") (display (tensor-get actual 0)) (newline)
(display "finite=") (display finite) (newline)
(display (if (near (tensor-get actual 0) finite 1e-4) "PASS" "FAIL"))
(newline)
```

Command:

```bash
eshkol-run --no-stdlib repro.esk -o repro
./repro
```

Observed output:

```text
reverse=0
finite=0.20413179971923154
FAIL
```

This blocks downstream use of pinned-core LayerNorm reverse AD. N2 will use a
separately reviewed explicit analytic backward provider rather than treating this
core result as usable or falling back to finite differences.

**Comment by tsotchke** [comment]
Confirmed, and the first case is worse on a current build than your report describes. You wrote that it "returns values that fail `[4, 3]`". What I get is:

```
(display (gradient (lambda (v) (+ (* (tensor-get v 0) (tensor-get v 0))
                                  (* 3.0 (tensor-get v 1))))
                   #(2.0 5.0)))
#(0 0)
```

A zero gradient, not a wrong-but-nonzero one. That matters for triage: a nonzero wrong value suggests an arithmetic error somewhere in the chain, while an exact `#(0 0)` for a function that plainly varies in both coordinates says the tangent never propagated at all — the derivative information is being dropped, not miscomputed. It also means the failure is invisible to any check that only asserts "the gradient is finite" or "training loss decreased", because a zero gradient simply stalls.

Your N2 addendum makes the same point from the other side: `layer-norm` returning a zero reverse gradient where the same compiled program's central difference is nonzero. Same signature, different operator. Including the finite-difference comparison in the reproducer is what makes it undeniable rather than a matter of opinion about the expected value.

**Status.** #556 addresses this and declares `Fixes #551` — forward-mode duals surviving layer-norm, scaled-dot attention and `tensor-get` on both engines. #532 (dense tensor AD node: matmul, reductions, elementwise and broadcast VJPs) is the structural work behind that area.

Both open, both currently **conflicting** with master and needing a rebase. That delay is ours. Thank you for the addendum on the pinned revision — re-reporting against the same SHA rather than a moved HEAD is what let this be checked directly.
